### ISSUE TITLE
'void' shouldn't be a silently indexable type

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#41021]
```ts
let x = [1, 2, 3, 4].forEach(x => { console.log(x));
x["hello"] = 123;
```

Expected: an error that `x` is possibly `void` or something
Actual: no errors

**Comment by DanielRosenwasser** [comment]
Actually, `never` seems to suffer from the same issue..

**Comment by ajafff** [comment]
Element access on `void` is already an error under `noImplicitAny`, only `never` seems to be indexable: https://www.typescriptlang.org/play?ts=4.1.3#code/CYUwxgNghgTiAEYD2A7AzgF3gNwFw6QEtgBuAWAChsBtAcgDMklaBdcqux5+AangFkoGABYA6GFBTAkAWwAUASjaVKoSLATJ0WFPhQhsIGOxScmrEkA
